filebeat/input/filestream/internal/input-logfile/store.go

```diff
@@ -294,6 +299,20 @@ func (s *store) writeState(r *resource) {
 	}
 }
 
+func (s *store) updateCursor(key string, cur interface{}) error {
+	r := s.ephemeralStore.Find(key, false)
+	if r == nil {
+		return fmt.Errorf("resource '%s' not found", key)
+	}
+	op, err := createUpdateOp(s, r, cur)
+	if err != nil {
+		return err
+	}
+	op.Execute(1)
```

Comment:
Update operations must be ordered, which means that update operations must only be applied after being confirmed by the ACK handler of the input. If you realy want to be able to update the cursor here, you must ensure that all pending update operations are discarded first.

possible solutions for implementing 'discard':
1. In the harvester ensure that the ack handler will drop pending updates
2. In the 'resource' add some kind of alive/version marker by pointer that is shared by all update operations. In order to unlink increase the version atomically and have the update check for the version before updating.

The version will give the 'resource' a new identity internally and you are in control of the cursor before starting a new harvester.

e.g.

```
type resource struct {
  version uint // current identity version. pending updates will be discarded when version is increased (pendingUpdates is set to 0, allowing the gc to pick up the resource and remove it early)

  ...
}

type updateOp struct {
  resourceVersion uint
  ...
}

func createUpdateOp(store *store, resource *resource, updates interface{}) (*updateOp, error) {
  ...
  op.resourceVersion = resource.version
  ...
}

func (op *updateOp) Execute(n uint) {
  ...
  if op.resourceVersion != op.resource.version {
    // resource is already 'new' and op has been unlinked. Do not write state to the persistent store
    return
  }

  ...
}
```